src/System.Linq/tests/SelectTests.cs

```diff
@@ -370,76 +370,6 @@ public void Select_SourceIsIEnumerable_ReturnsExpectedValues()
         }
 
         [Fact]
-        public void Select_SourceIsAnArray_CurrentIsDefaultOfTAfterEnumeration()
-        {
-            int[] source = new[] { 1 };
-            Func<int, int> selector = i => i + 1;
-
-            IEnumerable<int> query = source.Select(selector);
-
-            var enumerator = query.GetEnumerator();
-            while (enumerator.MoveNext()) ;
-
-            Assert.Equal(default(int), enumerator.Current);
-        }
-
-        [Fact]
-        public void Select_SourceIsAList_CurrentIsDefaultOfTAfterEnumeration()
-        {
-            List<int> source = new List<int>() { 1 };
-            Func<int, int> selector = i => i + 1;
-
-            IEnumerable<int> query = source.Select(selector);
-
-            var enumerator = query.GetEnumerator();
-            while (enumerator.MoveNext()) ;
-
-            Assert.Equal(default(int), enumerator.Current);
-        }
-
-        [Fact]
-        public void Select_SourceIsIReadOnlyCollection_CurrentIsDefaultOfTAfterEnumeration()
-        {
-            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int>() { 1 });
-            Func<int, int> selector = i => i + 1;
-
-            IEnumerable<int> query = source.Select(selector);
-
-            var enumerator = query.GetEnumerator();
-            while (enumerator.MoveNext()) ;
-
-            Assert.Equal(default(int), enumerator.Current);
-        }
-
-        [Fact]
-        public void Select_SourceIsICollection_CurrentIsDefaultOfTAfterEnumeration()
-        {
-            ICollection<int> source = new LinkedList<int>(new List<int>() { 1 });
-            Func<int, int> selector = i => i + 1;
-
-            IEnumerable<int> query = source.Select(selector);
-
-            var enumerator = query.GetEnumerator();
-            while (enumerator.MoveNext()) ;
-
-            Assert.Equal(default(int), enumerator.Current);
-        }
-
-        [Fact]
-        public void Select_SourceIsIEnumerable_CurrentIsDefaultOfTAfterEnumeration()
-        {
-            IEnumerable<int> source = Enumerable.Repeat(1, 1);
-            Func<int, int> selector = i => i + 1;
-
-            IEnumerable<int> query = source.Select(selector);
-
-            var enumerator = query.GetEnumerator();
-            while (enumerator.MoveNext()) ;
-
-            Assert.Equal(default(int), enumerator.Current);
-        }
```

Comment:
That's an interesting point. We've potentially introduced a very slight break in any cases we moved from `yield` to custom iterator. This introduces an equivalent slight break in the opposite direction. I'd be inclined to argue this should be somewhere where we feel free to change behaviour, but at the same time that the old behaviour is perhaps the preferable for the GC impact in cases where elements are large and the enumerable itself is held onto.